### PR TITLE
#740 #1580 Support multiple authentication schemes in one route

### LINKED ISSUE #740: Enhancement: Allow a reroute to specify more than one authentication scheme
##  New Feature
Allow a reroute to specify more than one authentication scheme for a reroute.  Specification could either be JSON array or comma separated.

## Motivation for New Feature
Sometimes more than one authentication scheme may be allowed for an API (e.g. JWT + Client Certificates). This feature would allow more than one scheme to be use on a single reroute eithout requiring a custom override of the authentication pipeline..

### LINKED ISSUE #1580: Support several AuthenticationProviderKey values in the Route
## New Feature
Opportunity to support several [AuthenticationProviderKey](https://ocelot.readthedocs.io/en/latest/features/authentication.html?highlight=AuthenticationProviderKey) values in the Route.

## Motivation for New Feature
At the moment, [Route](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Configuration/DownstreamRoute.cs) can contain one [AuthenticationProviderKey](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Configuration/AuthenticationOptions.cs#L12) value only. So, if we want to handle Route for different authentication providers, we have to specify two Routes and make Route path unique. Sometime it is not so easy to specify unique paths for the same endpoint.

## Useful Links
- [Features | Authentication](https://ocelot.readthedocs.io/en/latest/features/authentication.html) &rarr; ([authentication.rst](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/authentication.rst)), Ocelot docs
- [Authorize with a specific scheme in ASP.NET Core | Use multiple authentication schemes](https://learn.microsoft.com/en-us/aspnet/core/security/authorization/limitingidentitybyscheme?view=aspnetcore-8.0#use-multiple-authentication-schemes), Microsoft Learn
- [Basic Authentication in C# / .NET](https://scatteredcode.net/basic-authentication-in-c-sharp-net/), ScatteredCode by Alexandru Puiu
- [.NET 6.0 - Basic Authentication Tutorial with Example API](https://jasonwatmore.com/post/2021/12/20/net-6-basic-authentication-tutorial-with-example-api), Jason Watmore's Blog

## Related
- #446 
- #1002

### PR DESCRIPTION
**Pull request by MayorSheFF**
## Closes #1580 #740 
- #1580 
- #740 

## Proposed Changes
  - opportunity to have the several authentication provider keys in Routes what allows to authenticate requests via different providers

### CODE REVIEW

**Comment by MayorSheFF** [comment]
@raman-m,

That is a new PR. Is it OK for you?

**Comment by raman-m** [comment]
>@MayorSheFF commented [on Dec 28](https://github.com/ThreeMammals/Ocelot/pull/1870#issuecomment-1871089330)

Thank you very much!

File: src/Ocelot/Configuration/File/FileAuthenticationOptions.cs [hunk 1]
```diff
@@ -1,28 +1,31 @@
...
-            return sb.ToString();
-        }
+        public override string ToString() => new StringBuilder()
+            .Append($"{nameof(AuthenticationProviderKey)}:{AuthenticationProviderKey},{nameof(AuthenticationProviderKeys)}:[")
+            .AppendJoin(',', AuthenticationProviderKeys)
+            .Append("],")
+            .Append($"{nameof(AllowedScopes)}:[")
+            .AppendJoin(',', AllowedScopes)
+            .Append(']')
+            .ToString();
```
Comment (RaynaldM) [inline src/Ocelot/Configuration/File/FileAuthenticationOptions.cs:29]:
```suggestion
        public override string ToString() =>new StringBuilder()
        .Append($"{nameof(AuthenticationProviderKey)}:{AuthenticationProviderKey},")
        .Append($"{nameof(AuthenticationProviderKeys)}:[{string.Join(',', AuthenticationProviderKeys)}],")
        .Append($"{nameof(AllowedScopes)}:[{string.Join(',', AllowedScopes)}]")
        .ToString();

```

Comment (raman-m) [reply src/Ocelot/Configuration/File/FileAuthenticationOptions.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1870/commits/27869622919503ba4749c6de271a02cb6df08ab1 ✔️

**Review by RaynaldM — APPROVED** [review]

**Comment by raman-m** [comment]
@ggnaegi @RaynaldM @MayorSheFF 
I have [the concern](https://github.com/ThreeMammals/Ocelot/issues/1580#issuecomment-1911917306) regarding new `AuthenticationProviderKeys` property.
We can reuse old `AuthenticationProviderKey` property and as a developer I haven't update a lot of code lines.
Seems adding the new `AuthenticationProviderKeys` property is overhead design...
Please share your opinion!

**Comment by ggnaegi** [comment]
@raman-m I'm not a fan of the language approach with delimiters. You will need a parser for that, with some risks. I would recommend keeping the variant proposed by @MayorSheFF and mark the property AuthenticationProviderKey as deprecated.

**Comment by RaynaldM** [comment]
@raman-m I don't think I'm qualified to give an opinion (we don't use auth through the gateway).

**Comment by raman-m** [comment]
@ggnaegi Good, make sense... Everybody loves its own design... Seems it is to late to change design.

But the property is not deprecated. We still consider it as primary auth key, because of backward compatibility.
But what did you mean actually?
Marking the property with `[Obsolete]` attribute?

**Comment by ggnaegi** [comment]
> @ggnaegi Good, make sense... Everybody loves its own design... Seems it is to late to change design.
> 
> But the property is not deprecated. We still consider it as primary auth key, because of backward compatibility.
> But what did you mean actually?
> Marking the property with `[Obsolete]` attribute?

Yes mark it obsolete, so we promote the new approach...

**Comment by ggnaegi** [comment]
@MayorSheFF could you give me the permissions on your PR so I can finish the last steps? Thanks ggnaegi

**Comment by raman-m** [comment]
@ggnaegi  I have done feature design. And added major required acceptance tests 👉https://github.com/ThreeMammals/Ocelot/pull/1870/commits/0f0a856b44770d6d0dbfcc5c8c31445a0d40b5b8 related to Identity Server 4.
Could you review please? And if no objections then I'm going to merge.

**Review by ggnaegi — APPROVED** [review]

**Review by raman-m — APPROVED** [review]
Code Complete! 🟢 
- Code review ➕➕➕
- High quality source code ➕ 
- Unit tests ➕ 
- Acceptance tests ➕ 
- Updated Docs ➕ 
- Release notes ❔

**Comment by raman-m** [comment]
The [build](https://circleci.com/gh/ThreeMammals/Ocelot/3239) is 🟢 

Ready for delivery! ✔️

**Pull request merged**